pkg/kubernetes/client.js

```diff
@@ -763,6 +664,107 @@ define([
+                var scheme = schemes.shift();
+
+                /* No further ports to try? */
+                if (!scheme) {
+                    var ex = new Error(_("Couldn't find running kube-apiserver"));
+                    ex.problem = "not-found";
+                    dfd.reject(ex);
+                    return;
+                }
+
+                var http = cockpit.http(scheme.port, scheme);
+                var openshift = null;
+
+                /* A supplementary request to check if openshift */
+                aux = http.get(API_OPENSHIFT)
+                    .always(function() {
+                        openshift = (this.state() == "resolved");
+                    });
+
+                /* The main /api request */
+                req = http.get(API_KUBE)
+                    .done(function(data) {
+                        req = null;
+
+                        /*
+                         * We expect a response that looks something like:
+                         * { "versions": [ "v1beta1", "v1beta2", "v1" ] }
+                         */
+                        var response;
+                        try {
+                            response = JSON.parse(data);
+                        } catch(ex) {
+                            debug("not an api endpoint without JSON data on:", scheme);
+                            step();
+                            return;
+                        }
+                        if (response && response.versions) {
+                            aux.always(function() {
+                                response.flavor = openshift ? "openshift" : "kubernetes";
+                                dfd.resolve(http, response);
+                            });
+
+                        } else {
+                            debug("not a kube-apiserver endpoint on:", scheme);
+                            step();
+                        }
+                    })
+                    .fail(function(ex, response) {
+                        req = null;
+
+                        if (ex.problem === "not-found") {
+                            debug("api endpoint not found on:", scheme);
+                            step();
+                        } else {
+                            if (ex.problem !== "cancelled")
+                                debug("connecting to endpoint failed:", scheme, ex);
+                            dfd.reject(ex, response);
+                        }
+                    });
+            }
+
+            /* Load the kube config, and then try to start connecting */
+            cockpit.spawn(["/usr/bin/kubectl", "config", "view", "--output=json", "--raw"])
+                .fail(function(ex, output) {
+                    if (output)
+                        console.warn(output);
+                    else
+                        console.warn(ex);
+                })
+                .done(function(data) {
+                    self.config = data;
```

Comment:
How do we know we actually connected using this config?

I would suggest the following:
 1. Have config.parse_scheme() a 'kubeconfig' field to the scheme.
 2. Remove that field before passing the scheme to cockpit.http()

Include the kubeconfig when resolving the deferred. This way this whole function doesn't need to be nested.